### PR TITLE
Rewrite "email confirmation" + a few fixes (#308, #60, …)

### PR DESCRIPTION
**Pull request by jeandat**
# Pull Request Content

This PR started with a rewrite of the email confirmation workflow. Indeed, the “verify email” template was not internationalized and there was several issues (notably #308).
But it is now a bit larger (notably #60) and I’m sorry if its a bit rude of me. Maybe I should have done this in several PR.

Here is the content of the PR.

# Email confirmation component

- Email confirmation component is now internationalized via several inputs in auth component: `verifyEmailTitleText`, `verifyEmailConfirmationText`, `verifyEmailGoBackText`, `sendNewVerificationEmailText`, `signOutText`, `messageOnEmailConfirmationSuccess`

- This component is not rendered in the signup tab anymore. Instead it is rendered in place of all tabs cause once you are signed up and waiting email confirmation, there is no meaning in showing the sign-in/up forms.

- State stay coherent accross reloads ; meaning, if you reload the page and show again the auth component (either directly or via the guard), you come back to email confirmation component. This is the default behavior. Those who do no want that behaviour whatever the reason can opt-out via a new config parameter `guardProtectedRoutesUntilEmailIsVerified`.

- It is also possible to override completely the template used for email confirmation by providing your own template via a new input in auth component `verifyEmailTemplate`.

- Finally, there is three new action buttons that allows you to send again a confirmation email, signout, or pursue (continue) your navigation. The "continue" button will find the best appropriate next url by either using the "go back url" provided or if you came from the guard, the previously intercepted url (#60). Indeed, the guard now add a redirectUrl query parameter before redirecting to the fallback url. Before navigating, user info are reloaded in case email has been verified in the mean time.

# Auth component

- Once signed in, a new template is shown in place of all tabs that give you basic informations like display name, email, photo, etc. It was strange to keep the signin/up forms visible.
- There is also action buttons that allow you to either sign out or pursue your navigation. The "continue" button behaviour is identical to what has been described for the "verify" component. Regarding the `redirectUrl` query parameter, people can of course use that information in their `onSuccess` callback if they want to do the redirection themselves proactively after sign in.

- Hardcoded error messages related to passwords length have been internationalized via new inputs: `passwordErrorMinLengthText`, `passwordErrorMaxLengthText`.

- A new input `providersTheme` allow you to... wait for it... customize providers theme :)

- In forms, submit button is disabled until form is valid

- Providers are hidden when the "reset password" is active cause it seems to me it doesn't make sense to allow that kind of actions in that context.

- `messageOnAuthError` has evolve to accept also a `function` which allows you to instrument dynamically the error message based on error. If that function returns a falsy value, no toast is rendered. For instance in one of my project I'm providing that function:

```typescript
getErrorMessage = (error: { code: string }) => {
  if (NGX_AUTH_ERRORS_TO_IGNORE.includes(error.code)) return;
  const translationKey = NGX_AUTH_ERRORS[error.code] || 'core.genericError';
  return this.translateService.instant(translationKey);
};
```

- Toast messages now reuse current `MatSnackBarConfig` instead of hardcoding a duration. In case there is no duration configured (by default 0), an action button is added to permit the end-user to dismiss it.

# Chore

- I moved svg icons declaration in module cause it seems to me they should be declared as soon as possible once module is loaded.

- I deactivated tslint rule about max lengths (and several deprecated ones) cause I think this is the wrong way to enforce consistent style. I intend to submit another PR for that. Prettier should be configured for the whole project and executed every time a commit is done. Life would be much easier for people contributing and consistent code style would be enforced without effort. If you are interested I can do a WIP PR to discuss about this.

### CODE REVIEW

**Comment by kuncevic** [comment]
Can email verification be optional?

**Comment by jeandat** [comment]
Didn't test that but it may be transparent. What is the value of "emailVerified" field in a firebase User object if you disable email verification completely in your firebase account? It it is automatically true, it will probably work. I am in no situation to test that currently though.

**Comment by jeandat** [comment]
Hi @AnthonyNahas ,

I just saw that you are much active on the project and currently refactoring things.
IMHO, this PR add real value to the project. It is larger than just "email verification". It fixes several issues (some referenced and others not).

Would you mind reviewing it?
If you plan to integrate it, now would be the right moment before committing anything else cause it will not be possible anymore later if master diverge too much.

I spent some time reading and understanding your codebase trying to add value to it without introducing breaking changes. Hope you can give it some time :)

I would love for it to be merged :)
Have a good day.

**Comment by AnthonyNahas** [comment]
I will check that asap

File: src/module/components/ngx-auth-firebaseui/auth.component.ts [hunk 1]
```diff
@@ -36,14 +24,31 @@ export const PHONE_NUMBER_REGEX = new RegExp(/^\+(?:[0-9] ?){6,14}[0-9]$/);
...
+
+  // i18n translations to use in default template for email verification.
+  // See email-confirmation component
+  @Input() verifyEmailTitleText: string;
+  @Input() verifyEmailConfirmationText: string;
+  @Input() verifyEmailGoBackText: string;
+  @Input() sendNewVerificationEmailText: string;
+  @Input() signOutText: string;
+
+  @Input() providers: AuthProvider[] | AuthProvider; //  google, facebook, twitter, github as array or all as one single string
```
Comment (AnthonyNahas) [inline src/module/components/ngx-auth-firebaseui/auth.component.ts]:
```html 
@Input() providers: AuthProvider[] | AuthProvider; 
```

the above property input should be initialized --> 

```html 
@Input() providers: AuthProvider[] | AuthProvider = AuthProvider.ALL; 
```

File: src/module/components/ngx-auth-firebaseui/auth.component.ts [hunk 2]
```diff
@@ -36,14 +24,31 @@ export const PHONE_NUMBER_REGEX = new RegExp(/^\+(?:[0-9] ?){6,14}[0-9]$/);
 @Component({
   selector: 'ngx-auth-firebaseui',
   templateUrl: 'auth.component.html',
-  styleUrls: ['auth.component.scss']
+  styleUrls: ['auth.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
```
Comment (AnthonyNahas) [inline src/module/components/ngx-auth-firebaseui/auth.component.ts]: why did you change the ChangeDetectionStrategy ?

**Comment by AnthonyNahas** [comment]
@jeandat first of all, thank you for the great work!
I spent like more than 2h trying to merge your PR smoothly and I am almost done with that! 
We made some changes at the same time for some main features of the library! 
There were some minor bugs that I have just fixed them.. and enhanced the code a little bit.
I've added animations to every component and much more.

However, do you think the next update should be a major or minor update?
Do you think that this PR has any breaking changes?

Thank you one more time for the contribution 👍 
The final merge to the master branch will be very soon

cheers 🍻

**Comment by AnthonyNahas** [comment]
congrats 🎆
